integr8ly/integreatly-operator
# RHOAM OSD 4.12 support

# Issue link
https://issues.redhat.com/browse/MGDAPI-4810

# What
- Dependencies bump to support osd 4.12
- Marin3r bump to support 4.12
- Envoy image bump to support marin3r bump
- Keycloak client bump to support 4.12
- 3scale Operator and APIs bump to support 4.12 (in progress - no new 3scale version available as of yet)
- Observability Operator and APIs bump to support 4.12 (in progress - no new OO available as of yet)
- Other changes around controllers, tests and other various files:
   - I would appreciate running through RHMI Controller refactor and 3scale reconciler refactors in particular

# Verification steps
Due to the nature of this PR there has been a need to update some code logic not directly relevant to OSD 4.12 but caused by it. To simplify the verification process the changes have been included in this PR rather than in separate ones.

NOTE: Still waiting for new OO and 3scale Operator version.

## Verification of rate limiting refferers header
- Install RHOAM from this branch either via OLM or local, with either cluster storage set to true or false, whichever is easier for you
- Once RHOAM is installed navigate to 3scale admin portal. You can get routes from 3scale namespace and login creds from system-seed secret
- Once in the 3scale portal navigate to the default product Integration/Policies - click on product > integration > policies
- Add policy chain - 3scale Referrer and update (bottom right)
- Next, go to Application\Settings\Usage Rules and set “Required referrer filtering” to true/selected and update (bottom right button)
- Next, click on Applications > Listings > Developers app and add 2 random Referrers Filters - "dummy" and "localhost"
- Next, go back to the main view of the product and go to Integration > Configuration and promote your product to v2 to stage env
-  Copy curl command
- In your terminal paste the curl command copied to ping apicast but add the following to it:
`curl -i -H "Referer: localho" "https://api-3scale-apicast..../?user_key=<YOUR_USER_KEY>"`
Run the command and the output here should be "Authentication failed"
- Now fix the referer value to either localhost or dummy and curl again. The command should succeed.
- Next, scale down rhoam operator and navigate to marin3r namespace. Find the rate limiting config map and edit the maxValue to 5. Save changes and rollout deployment of the marin3r pods. 
- After a minute (once deployments of marin3r pods is fully ready) curl again with correct Referer and confirm that the requests are 429 after 5 successful curls.
